tc-messaging/src/main/java/com/tc/net/protocol/transport/WireProtocolGroupMessageImpl.java

```diff
@@ -49,115 +49,104 @@
-    TCByteBuffer[] msgs = messagePayloadByteBuffers;
-
-    if (msgs.length > 1) {
-      fullMsgsBytes = new byte[((WireProtocolHeader) getHeader()).getTotalPacketLength()
-                               - ((WireProtocolHeader) getHeader()).getHeaderByteLength()];
-      int copyPos = 0;
-      for (int i = 0; i < msgs.length; i++) {
-        System.arraycopy(msgs[i].array(), 0, fullMsgsBytes, copyPos, msgs[i].limit());
-        copyPos += msgs[i].limit();
+  @Override
+  public boolean prepareToSend() {
+      setPayload(generatePayload());
+      getWireProtocolHeader().setMessageCount(messagePayloads.size());
+      getWireProtocolHeader().finalizeHeader(getTotalLength());
+      return getWireProtocolHeader().getMessageCount() > 0;
+  }
+  
+  private TCByteBuffer[] generatePayload() {
+    List<TCByteBuffer> msgs = new ArrayList<>(messagePayloads.size() * 2);
+    Iterator<TCActionNetworkMessage> msgI = messagePayloads.iterator();
+    while (msgI.hasNext()) {
+      TCActionNetworkMessage msg = msgI.next();
+      if (msg.commit()) {
+        TCByteBuffer tcb = TCByteBufferFactory.getInstance((Integer.SIZE + Short.SIZE) / 8);
+        tcb.putInt(msg.getTotalLength());
+        tcb.putShort(WireProtocolHeader.getProtocolForMessageClass(msg));
+        tcb.flip();
+
+        msgs.add(tcb);
+        // referring to the original payload buffers
+        msgs.addAll(Arrays.asList(msg.getEntireMessageData()));
+      } else {
+        msg.complete();
+        msgI.remove();
       }
-    } else {
-      fullMsgsBytes = msgs[0].array();
     }
-
-    TCByteBuffer b = TCByteBufferFactory.wrap(fullMsgsBytes);
-    for (int i = 0; i < getTotalMessageCount(); i++) {
-      int msgLen = b.getInt();
-      short msgProto = b.getShort();
-
-      // XXX: we are giving out 4K BB for a smaller msgs too; Though it is recycled, can do some opti., here
-      TCByteBuffer[] bufs = new TCByteBuffer[] {TCByteBufferFactory.getInstance(msgLen)};
-
-              
-      for (TCByteBuffer buf : bufs) {
-        b.get(buf.array(), 0, buf.limit());
+    return msgs.toArray(new TCByteBuffer[msgs.size()]);
+  }
+  
+  private List<TCNetworkMessage> getMessagesFromByteBuffers() {
+    ArrayList<TCNetworkMessage> messages = new ArrayList<>();
+
+    TCByteBufferInputStream msgs = new TCByteBufferInputStream(getPayload());
+
+    for (int i = 0; i < getWireProtocolHeader().getMessageCount(); i++) {
+      try {
+        int msgLen = msgs.readInt();
+        short msgProto = msgs.readShort();
+        
+        WireProtocolHeader hdr;
+        hdr = (WireProtocolHeader) getWireProtocolHeader().clone();
+        hdr.setTotalPacketLength(hdr.getHeaderByteLength() + msgLen);
+        hdr.setProtocol(msgProto);
+        hdr.setMessageCount(1);
+        hdr.computeChecksum();
+        WireProtocolMessage msg = new WireProtocolMessageImpl(this.sourceConnection, hdr, new TCByteBuffer[] {msgs.read(msgLen)});
+        messages.add(msg);
+      } catch (IOException ioe) {
```

Comment:
This catch and ignore seems dangerous. What we probably should be doing is reworking TCByteBufferInputStream to only throw `EOFException` since we know IOException is impossible. That should probably then be caught here and rethrown as some kind of protocol error that will get handled sanely?